tls/x509utils/pair.go

```diff
@@ -1,18 +1,66 @@
 package x509utils
 
-import "crypto/x509"
+import (
+	"bytes"
+	"crypto"
+	"crypto/x509"
+)
+
+// PrivateKeyEqual tells if two private keys are the same.
+// nil keys aren't considered comparable.
+func PrivateKeyEqual(a, b crypto.PrivateKey) bool {
+	if a2, ok := a.(PrivateKey); ok {
+		return a2.Equal(b)
+	}
+	return false
+}
+
+// PublicKeyEqual tells if two public keys are the same.
+// nil keys aren't considered comparable.
+func PublicKeyEqual(a, b crypto.PublicKey) bool {
+	if a2, ok := a.(PublicKey); ok {
+		return a2.Equal(b)
+	}
+	return false
+}
 
 // ValidCertKeyPair confirms the given key can use the given certificate.
-func ValidCertKeyPair(cert *x509.Certificate, key PrivateKey) bool {
+// nil keys aren't considered comparable.
+func ValidCertKeyPair(cert *x509.Certificate, key crypto.PrivateKey) bool {
 	if cert == nil || key == nil {
-		// missing
 		return false
 	}
-	pub, ok := cert.PublicKey.(PublicKey)
-	if !ok {
-		// invalid - unreachable
+
+	if pk, ok := key.(PrivateKey); ok {
+		return PublicKeyEqual(cert.PublicKey, pk.Public())
+	}
+
+	return false
+}
+
+// ValidKeyPair confirms the public key matches the private one.
+// nil keys aren't considered comparable.
+func ValidKeyPair(pub crypto.PublicKey, key crypto.PrivateKey) bool {
```

Comment:
`func MatchingKeys(pub crypto.PublicKey, key crypto.PrivateKey) bool {` ?

Reply:
> `func MatchingKeys(pub crypto.PublicKey, key crypto.PrivateKey) bool {` ?

why do you think this doesn't fit the `ValidFoo(...) bool` pattern?

Reply:
In my mind `Valid` certifies that something is correct, valid, flawless. `Match` means that it is like something else, similar to it, indistinguishable from it. This is more a philosophic observation. If a public key matches the private one does it make it valid? Or is it just computable from the private one?

Reply:
> In my mind `Valid` certifies that something is correct, valid, flawless. `Match` means that it is like something else, similar to it, indistinguishable from it. This is more a philosophic observation. If a public key matches the private one does it make it valid? Or is it just computable from the private one?

I think that next step falls under the `Verify(....) error` pattern

Reply:
I mean, we are using a weaker test on `ValidCertKeyPair()`. Only comparing public keys match, not even the `notBefore`/`notAfter`.

Reply:
On second thoughts `ValidKeyPair` is ok, if you read it carefully it is only saying that the `KeyPair` is valid, as in it is a pair.